File: core/src/ru/myitschool/cubegame/dungeon/Dungeon.java
package ru.myitschool.cubegame.dungeon;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.cubegame.ai.pathfinding.GraphStorage;
import ru.myitschool.cubegame.ai.pathfinding.Node;
import ru.myitschool.cubegame.entities.Character;
import ru.myitschool.cubegame.story.quest.Quest;

import java.awt.*;

/**
 * Created by Voyager on 25.03.2018.
 */
public class Dungeon {


    public static final int ROOM_WIDTH = 8;
    public static final int ROOM_HEIGHT = 8;

    private Quest quest;

    int placedRoomsCount = 0;

    private Array<Vector2> charPoses = new Array<>();

    private Array<Room> roomsPool = new Array<Room>();
    private Array<Room> roomsPlaced = new Array<Room>();

    public Dungeon(Quest quest) {
        this(quest, getDefaultCharPoses());
    }

    public Dungeon(Quest quest, Array<Vector2> charPoses) {
        this.quest = quest;
        this.charPoses = charPoses;
        createRooms(quest.getRoomsMax());
        placeChars();
        Exit.resetExitsCount();
        Exit.setExitsMax(quest.getRoomsMax() - 1);
    }

    private static Array<Vector2> getDefaultCharPoses(){
        Array<Vector2> array = new Array<>();
        array.add(new Vector2(3, 3));
        array.add(new Vector2(4, 3));
        array.add(new Vector2(4, 4));
        array.add(new Vector2(3, 4));
        return array;
    }

    public Quest getQuest() {
        return quest;
    }

    public int getRoomPos() {
        return getQuest().getRoomPos();
    }

    public int getRoomCountMax() {
        return getQuest().getRoomsMax();
    }

    public Array<Room> getRoomsPool() {
        return roomsPool;
    }

    public Array<Room> getRoomsPlaced() {
        return roomsPlaced;
    }

    public void moveRooms(Vector2 move){
        for (Room room : roomsPlaced){
            room.changePos(move);
        }
    }

    public Room getPlacedRoom(int roomX, int roomY){
        for (Room room : roomsPlaced){
            if (room.getX() == roomX && room.getY() == roomY){
                return room;
            }
        }
        return null;
    }

    public Room popRoom(){
        Room room = roomsPool.get(0);
        roomsPool.removeIndex(0);
        roomsPlaced.add(room);
        placedRoomsCount++;
        return room;
    }

    public Room placeRoom(int x, int y, int side){
        Room room = popRoom();
        room.rotate(side);

        int roomX = x / ROOM_WIDTH;
        int roomY = y / ROOM_HEIGHT;
        System.out.println(roomX + ",ggnbgn " + roomY);
        room.setPos(roomX, roomY);
        Array<Integer> sides = new Array<Integer>();

        Room left = getPlacedRoom(roomX - 1, roomY);
        Room right = getPlacedRoom(roomX + 1, roomY);
        Room up = getPlacedRoom(roomX, roomY - 1);
        Room down = getPlacedRoom(roomX, roomY + 1);

        boolean leftExit = false;
        boolean rightExit = false;
        boolean topExit = false;
        boolean bottomExit = false;

        if (left != null) {
            if (left.hasExit(Exit.DIRECTION_EAST)) {
                System.out.println("Left: Yes");
                leftExit = true;
                room.addDoor(Exit.DIRECTION_WEST);
            }
        } else {
            sides.add(Exit.DIRECTION_WEST);
        }
        if (right != null){
            if (right.hasExit(Exit.DIRECTION_WEST)){
                System.out.println("Right: Yes");
                rightExit = true;
                room.addDoor(Exit.DIRECTION_EAST);
            }
        } else {
            sides.add(Exit.DIRECTION_EAST);
        }
        if (up != null){
            if (up.hasExit(Exit.DIRECTION_SOUTH)){
                System.out.println("Up: Yes");
                topExit = true;
                room.addDoor(Exit.DIRECTION_NORTH);
            }
        } else {
            sides.add(Exit.DIRECTION_NORTH);
        }
        if (down != null){
            if (down.hasExit(Exit.DIRECTION_NORTH)){
                System.out.println("Down: Yes");
                bottomExit = true;
                room.addDoor(Exit.DIRECTION_SOUTH);
            }
        } else {
            sides.add(Exit.DIRECTION_SOUTH);
        }
        room.addExits(sides);

        GraphStorage.addTopNode(new Node(roomX, roomY), leftExit, rightExit, topExit, bottomExit);
        room.complete();

        for (Exit exit : room.getExits()){
            Point[] cells = exit.getExitCells();
            for (Point cell : cells){
                cell.x += x;
                cell.y += y;
            }
            new Door(exit.getDirection(), cells);
        }

        return room;
    }

    private void createRooms(int count){
        int exitRoom = getRoomPos();
        System.out.println(exitRoom + " - exit room");
        for (int i = 0; i < count; i++) {
            Room room;
            if (i == 0) {
                room = createRoom(true);
            } else {
                room = createRoom(false);
                if (i == exitRoom - 1) {
                    room = quest.exitRoomConstruction(room);
                }
            }
            roomsPool.add(room);
        }
    }

    private Room createRoom(boolean first){
        Room room;
        if (first){
            room = new Room("rooms/default.room");
            room.setMobs(false);
        } else {
            room = new Room("rooms/corridor.room");
        }
        return room;
    }

    public void placeChars(){
        for (int i = 0; i < Character.getChars().size; i++) {
            Character character = Character.getChar(i);
            Vector2 pos = charPoses.get(i);
            character.setCellPos(pos);
        }
    }
}

File: core/src/com/wolg_vlad/dcrawler/dungeon/Dungeon.kt
package com.wolg_vlad.dcrawler.dungeon

import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.Character
import com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph.GraphStorage
import com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph.RoomNode
import com.wolg_vlad.dcrawler.story.quest.Quest
import com.wolg_vlad.dcrawler.utils.SeededRandom

/**
 * Created by Voyager on 25.03.2018.
 */
class Dungeon @JvmOverloads constructor(val quest: Quest, private val charPoses: Array<Vector2> = defaultCharPoses) {
    var placedRoomsCount = 0
    val roomsPool = Array<Room>()
    val roomsPlaced = Array<Room>()

    val endRoom: Int
        get() = quest.endRoom

    val roomCountMax: Int
        get() = quest.roomsMax

    fun moveRooms(move: Vector2) {
        for (room in roomsPlaced) {
            room.changePos(move)
        }
    }

    fun getPlacedRoom(roomX: Int, roomY: Int): Room? {
        for (room in roomsPlaced) {
            if (room.x == roomX && room.y == roomY) {
                return room
            }
        }
        return null
    }

    fun popRoom(): Room {
        val room = roomsPool[0]
        roomsPool.removeIndex(0)
        roomsPlaced.add(room)
        placedRoomsCount++
        return room
    }

    fun placeRoom(x: Int, y: Int, side: Int): Room {
        val room = popRoom()
        room.rotate(side)
        val roomX = x / ROOM_WIDTH
        val roomY = y / ROOM_HEIGHT
        println("$roomX,ggnbgn $roomY")
        room.setPos(roomX, roomY)
        val sides = Array<Int>()
        val left = getPlacedRoom(roomX - 1, roomY)
        val right = getPlacedRoom(roomX + 1, roomY)
        val up = getPlacedRoom(roomX, roomY - 1)
        val down = getPlacedRoom(roomX, roomY + 1)
        var leftExit = false
        var rightExit = false
        var topExit = false
        var bottomExit = false
        if (left != null) {
            if (left.hasExit(Exit.DIRECTION_EAST)) {
                println("Left: Yes")
                leftExit = true
                room.addDoor(Exit.DIRECTION_WEST)
                val doorPoint = left.getExit(Exit.DIRECTION_EAST)!!.exitCells[0]
                println("Door removed! Cords: (" + doorPoint!!.x + "; " + doorPoint.y + ")")
                Door.removeDoor(doorPoint.x, doorPoint.y)
            }
        } else {
            sides.add(Exit.DIRECTION_WEST)
        }
        if (right != null) {
            if (right.hasExit(Exit.DIRECTION_WEST)) {
                println("Right: Yes")
                rightExit = true
                room.addDoor(Exit.DIRECTION_EAST)
                val doorPoint = right.getExit(Exit.DIRECTION_WEST)!!.exitCells[0]
                println("Door removed! Cords: (" + doorPoint!!.x + "; " + doorPoint.y + ")")
                Door.removeDoor(doorPoint.x, doorPoint.y)
            }
        } else {
            sides.add(Exit.DIRECTION_EAST)
        }
        if (up != null) {
            if (up.hasExit(Exit.DIRECTION_SOUTH)) {
                println("Up: Yes")
                topExit = true
                room.addDoor(Exit.DIRECTION_NORTH)
                val doorPoint = up.getExit(Exit.DIRECTION_SOUTH)!!.exitCells[0]
                println("Door removed! Cords: (" + doorPoint!!.x + "; " + doorPoint.y + ")")
                Door.removeDoor(doorPoint.x, doorPoint.y)
            }
        } else {
            sides.add(Exit.DIRECTION_NORTH)
        }
        if (down != null) {
            if (down.hasExit(Exit.DIRECTION_NORTH)) {
                println("Down: Yes")
                bottomExit = true
                room.addDoor(Exit.DIRECTION_SOUTH)
                val doorPoint = down.getExit(Exit.DIRECTION_NORTH)!!.exitCells[0]
                println("Door removed! Cords: (" + doorPoint!!.x + "; " + doorPoint.y + ")")
                Door.removeDoor(doorPoint.x, doorPoint.y)
            }
        } else {
            sides.add(Exit.DIRECTION_SOUTH)
        }
        room.addExits(sides)
        GraphStorage.addTopNode(RoomNode(roomX, roomY, room))
        room.complete()
        for (exit in room.exits) {
            val cells = exit.exitCells
            for (cell in cells!!) {
                cell!!.x += x
                cell!!.y += y
            }
            Door(exit.direction, cells)
        }
        return room
    }

    private fun createRooms(count: Int) {
        val exitRoom = endRoom
        println("$exitRoom - exit room")
        for (i in 0 until count) {
            var room: Room
            if (i == 0) {
                room = createRoom(true)
            } else {
                room = createRoom(false)
                if (i == exitRoom - 1) {
                    room = quest.exitRoomConstruction(room)
                }
            }
            roomsPool.add(room)
        }
    }

    private fun createRoom(first: Boolean): Room {
        val room: Room
        if (first) {
            room = Room("rooms/default.room")
            room.setMobs(false)
        } else {
            val rnd = SeededRandom.getInstance().nextInt(10)
            room = if (rnd > 3) {
                Room("rooms/corridor.room")
            } else {
                Room("rooms/default.room")
            }
        }
        return room
    }

    fun placeChars() {
        for (i in 0 until Character.chars.size) {
            val character = Character.getChar(i)
            val pos = charPoses[i]
            character.setCellPos(pos)
        }
    }

    companion object {
        const val ROOM_WIDTH = 8
        const val ROOM_HEIGHT = 8
        private val defaultCharPoses: Array<Vector2>
            private get() {
                val array = Array<Vector2>()
                array.add(Vector2(3F, 3F))
                array.add(Vector2(4F, 3F))
                array.add(Vector2(4F, 4F))
                array.add(Vector2(3F, 4F))
                return array
            }
    }

    init {
        createRooms(quest.roomsMax)
        placeChars()
        Exit.resetExitsCount()
        Exit.exitsMax = quest.roomsMax - 1
    }
}